File: sliding-window & 2 pointers/239.js

//  Tc = O(nK) 
//  Sc = O(k)
// brute force 
function bruteForceMaxSlidingWindow(nums, k){
    const result = []
    for(let i = 0 ; i<nums.length; i++){
        result.push(Math.max(...nums.slice(i, i+k)))
    }
    return result
}

//  Tc = O(n) 
//  Sc = O(k)
// Monotonic Deque

function maxSlidingWindow(nums, k) {
    // Initialize a deque (will use array to simulate deque behavior)
    const deq = [];
    const result = [];
    
    for (let i = 0; i < nums.length; i++) {
        // Remove indices from front that are outside current window (i - k + 1 is window start)
        while (deq.length > 0 && deq[0] <= i - k) {
            deq.shift(); // Remove from front
        }
        
        // Remove indices from back whose values are <= current number
        // This maintains deque in decreasing order
        while (deq.length > 0 && nums[deq[deq.length - 1]] <= nums[i]) {
            deq.pop(); // Remove from back
        }
        
        // Add current index to deque
        deq.push(i);
        
        // Once we've processed at least k elements, start recording results
        if (i >= k - 1) {
            // Front of deque always has max for current window
            result.push(nums[deq[0]]); 
        }
    }
    
    return result;
}

// Example usage
const nums = [1, 3, -1, -3, 5, 3, 6, 7];
const k = 3;
console.log(maxSlidingWindow(nums, k)); // Output: [3, 3, 5, 5, 6, 7]

File: sliding-window & 2 pointers/239.py

# Tc = O(n) for each element exactly one push and pop
# Sc = O(k)
from collections import deque
def maxSlidingWindow(nums,k):
    deq = deque()
    result = []
    for i, num in enumerate(nums):
        while deq and deq[0] <= i - k:
            deq.popleft()
        while deq and nums[deq[-1]] < num:
            deq.pop()
        deq.append(i)
        if i >= k - 1:
            result.append(nums[deq[0]])
    return result

from collections import deque

def maxSlidingWindow(nums, k):
    
    # Initialize a deque to store indices of elements in the current window
    # The deque maintains elements in decreasing order of their values
    deq = deque()
    result = []
    
    # Iterate through each element in the input array
    for i, num in enumerate(nums):
        # Remove indices from the front that are outside the current window
        # Window boundaries: [i-k+1, i]
        while deq and deq[0] <= i - k:
            deq.popleft()
        
        # Remove indices from the back whose corresponding elements are <= current element
        # This maintains the deque in decreasing order
        while deq and nums[deq[-1]] <= num:
            deq.pop()
        
        # Add current element's index to the deque
        deq.append(i)
        
        # Once window size is reached, add the maximum to the result
        # The maximum is always at the front of the deque
        if i >= k - 1:
            result.append(nums[deq[0]])

    return result




